docs/proposals/31-inplace-node-update.md

```diff
@@ -0,0 +1,389 @@
+
+#### Roles in Failure Handling
+
+**`gardener-node-agent`**
+`gardener-node-agent` is responsible for initiating updates, with specific timeouts set for Kubernetes and machine image updates. When an error occurs it will be classified as follows:
+
+- **Retriable Error**: Temporary errors, such as network interruptions, prompt `gardener-node-agent` to retry the update after a delay. Retriable error will also get classified as non-retriable error after a certain number of retries and `gardener-node-agent` labels the node with an `update-failed` label.
+- **Non-Retriable Error**: For critical issues, such as incompatible OS versions, `gardener-node-agent` labels the node with an `update-failed` label.
+
+**`machine-controller-manager`**
+`machine-controller-manager` prepares nodes for updates by adding the `ready-to-update` label and waits for `gardener-node-agent` to perform the update. `machine-controller-manager` monitors for status labels applied by `gardener-node-agent`:
+
+- **Timeout Handling**: After the `machine-controller-manager` adds the `ready-for-update` label to a node, it waits for the `gardener-node-agent` to complete the update within the timeout configured in `machine-controller-manager`. If `gardener-node-agent` does not complete the update within this timeout, `machine-controller-manager` will retry the rollout after a specified delay. The timeout on the `machine-controller-manager` side is essential for handling cases where `gardener-node-agent` fails to report any status, such as during a total failure of the OS update. If the update is unsuccessful after a certain number of retries, `machine-controller-manager` will add the `update-failed` label to the node.
+- **Failed Update Handling**: If a node is labeled with `update-failed`, `machine-controller-manager` stops further updates. The machine remains with the old `MachineSet`, and no additional nodes in that `MachineDeployment` will be updated until the node with `update-failed` label is fixed. The failure is propagated to the shoot status, notifying the operator or user, who may need to investigate the issue manually.
+  To resolve the issue, the error must be fixed manually, and the `update-failed` label should be removed from the node. This allows `machine-controller-manager` to proceed with the update process.
+
+> [!NOTE]
+> All mentioned timeouts and retries are subject to change, with exact values to be decided during implementation.
+
+### Additional Validations
+
+In addition to Kubernetes or Machine image updates, several other factors can trigger the rolling updates of nodes in the shoot clusters. These include:
+
+1. Modification of the following fields in the Shoot specification:
+   - `.spec.provider.workers[].machine.type`
+   - `.spec.provider.workers[].volume`
+   - `.spec.provider.workers[].cri`
+   - `.spec.systemComponents.nodeLocalDNS`
+2. Rotation of the `CA` or `ServiceAccount` signing key.
```

Comment:
How will you facilitate the needed updates to kubelet to get the new credentials?